compiler/optimizations/gpuTransforms.cpp

```diff
@@ -421,11 +427,55 @@ bool GpuizableLoop::determineIfShouldErrorIfNotGpuizable() {
   for_alist(expr, cfl->body) {
     CallExpr *call = toCallExpr(expr);
     if (call && call->isPrimitive(PRIM_ASSERT_ON_GPU)) {
-      return true;
+      return call;
     }
   }
 
-  return false;
+  return nullptr;
+}
+
+static const char* findImmediateStringValue(Expr* expr) {
```

Comment:
I agree that this shouldn't happen here. Before how we can achieve this in a better way, is there a good use case for this feature? If I am understanding right you can do something like `@assertGpuEligible("wasn't eligible because I didn't like it")` and this will override the compiler's generic message? Can user ever say something relevant here?

I agree that the compiler messages should be more descriptive, but it is only the compiler who can know about this isn't it?

Reply:
The feature currently makes use of this -- to be able to print `Loop contains @assertGpuEligible` vs `Loop contains assertOnGpu()`. As far as I can tell there's no better way, and GPU transformations happen so late that the tree is quite mangled. I do have a fallback value for if the immediate value couldn't be found

Reply:
Especially following the other conversation where we might just rename this to `@assertOnGpu` and keep the behavior of this PR, and the current standalone function. I think we can just remove/deprecate the standalone `assertOnGpu()`. In which case, this'll be moot, right?

Reply:
I still think I'd rather have the error distinguish between `@assertOnGpu` and `assertOnGpu()`. If you don't think that's necessary, I can remove the check, string argument to ASSERT_ON_GPU, etc.

Reply:
If we are removing `assertOnGpu()`, which is what I am proposing. There's nothing to distinguish.

If we are keeping the function, maybe the way to distinguish is to have the primitive take a bool that's `true` for attribute and `false` otherwise. Then, you can sort of hardcode the string in here.